machine/util/callbacks/early_stopping_callback.py

```diff
@@ -0,0 +1,94 @@
+from machine.util.callbacks import Callback
+
+
+class EarlyStoppingCallback(Callback):
+    """
+    Original callback taken from https://github.com/ncullen93/torchsample
+    Early Stopping to terminate training early under certain conditions
+    """
+
+    def __init__(self,
+                 monitor='eval_losses',
+                 lm_name=None,
+                 min_delta=0,
+                 patience=5,
+                 minimize=True):
+        """
+        EarlyStopping callback to exit the training loop if training or
+        validation loss or metric does not improve by a certain amount
+        for a certain number of epochs
+        Args:
+            monitor : string in {'eval_losses', 'eval_metrics', 'train_losses', 'train_metrics'}
+                whether to monitor train or val loss
+            lm_name: loss or metric name eg. 'Avg NLLoss' or 'Word Accuracy'
+                    If not specified then the first element
+                    in the monitor array is used
+                    (default None)
+            min_delta : float
+                minimum change in monitored value to qualify as improvement.
+                This number should be positive.
+            patience : integer
+                number of epochs to wait for improvment before terminating.
+                the counter be reset after each improvment
+            minimize: minimize quantity, if false then early stopping will maximize
+        """
+        if 'loss' in monitor:
+            self.loss = True
+        elif 'metric' in monitor:
+            self.loss = False
+        else:
+            raise ValueError(
+                "Monitor must be string in {'eval_losses', \
+                'eval_metrics', 'train_losses', 'train_metrics'}")
+
+        self.monitor = monitor
+        self.lm_name = lm_name
+        self.minimize = minimize
+
+        self.min_delta = min_delta
+        self.patience = patience
+        self.wait = 0
+        self.best_lm = 1e-15
+        super(EarlyStoppingCallback, self).__init__()
+
+    def on_train_begin(self, info=None):
+        self.wait = 0
+        if self.minimize:
+            self.best_lm = 1e15
```

Comment:
Just out of curiosity: is it preferable to set this to a very small/large number instead of just setting it to inf?

Reply:
I just took a look at Keras implementation and they use inf so I'll change that as well, i picked 1e15 arbitrarily at the beginning because I saw another callback implementation that did that.

Reply:
either is fine by me :) I was just curious!